### PR TITLE
TASK: Add upgrade to deduplicate parallel base workspace changes 

### PR DESCRIPTION
**Pull request by mhsdesign**
based on #5858 

Upgrade to deduplicate parallel base workspace changes

for bug https://github.com/neos/neos-development-collection/issues/5877

Workspace operations, also ChangeBaseWorkspace were not thread safe before the addition of workspace versioning
in the read model and thus safe soft constraint checks.
That resulted in the possibility that a single workspace was changed two or more times in parallel by the same user.
Because each ChangeBaseWorkspace can be slow and cleans up at the end the content stream via ContentStreamWasRemoved,
that results in possibly multiple illegal ContentStreamWasRemoved events and in total a fully illegal ChangeBaseWorkspace.

The upgrade first identifies any duplicate ContentStreamWasRemoved events on a single stream.
If found we assume check via their unique prefixed correlation id that they belong to a ChangeBaseWorkspace sequence.
Then we find all events of the ChangeBaseWorkspace sequences and that occurred during these the content stream removals.
If there ar no other concurrent changes on our workspace - which would have been illegal as well - and it's truly
the race condition as understood with only the events a ChangeBaseWorkspace emits we continue.
We identify the last and thus valid ChangeBaseWorkspace sequence and delete all events from the previous illegal sequence(s).
The ChangeBaseWorkspace sequences can even be interlaced due to the race condition but the correlation id identifies the last sequence to keep uniquely.

After the migration is applied the workspace will have the same new base workspace as without the migration,
but we deduplicated any temporary changes that happened in a race condition. No content on the workspaces is affected.

for further insights, please see behaviour tests `EventsDeduplicateBaseWorkspaceChangesUpgrade` as to what cruelties we must handle

TODO:
- [x] decide on 9.0 or 9.2 to target. -> 9.2 because the problem can still appear in 9.0 again and again
- [x] Naming of CLI command
- [x] Message

### CODE REVIEW

File: Neos.ContentRepository.Export/Tests/Behavior/Features/Bootstrap/FeatureContext.php [hunk 1]
```diff
@@ -16,8 +16,7 @@
 use Behat\Behat\Context\Context as BehatContext;
 use Behat\Behat\Hook\Scope\BeforeScenarioScope;
 use Neos\Behat\FlowBootstrapTrait;
-use Neos\ContentGraph\DoctrineDbalAdapter\Tests\Behavior\Features\Bootstrap\CrImportExportTrait;
-use Neos\ContentRepository\Core\ContentRepository;
+use Neos\ContentGraph\DoctrineDbalAdapter\Tests\Behavior\Features\Bootstrap\CrImportExportTrait;use Neos\ContentRepository\Core\ContentRepository;
```
Comment (bwaidelich) [inline Neos.ContentRepository.Export/Tests/Behavior/Features/Bootstrap/FeatureContext.php]: ?

File: Neos.ContentRepositoryRegistry/Classes/Upgrade/Command/CRUpgradeCommandController.php [hunk 2]
```diff
@@ -82,4 +83,42 @@ public function eventsRecordedAtToUtcCommand(string $contentRepository = 'defaul
             force: $force
         );
     }
+
+    /**
+     * Migration to deduplicate consecutive base workspace changes
+     *
+     * https://github.com/neos/neos-development-collection/issues/5877
+     *
+     * TODO
```
Comment (bwaidelich) [inline Neos.ContentRepositoryRegistry/Classes/Upgrade/Command/CRUpgradeCommandController.php]: TODO :)

File: Neos.ContentRepositoryRegistry/Classes/Upgrade/Shared/EventEnvelopeFactory.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+use Neos\EventStore\Model\Event\EventType;
+use Neos\EventStore\Model\Event\SequenceNumber;
+use Neos\EventStore\Model\Event\StreamName;
+use Neos\EventStore\Model\Event\Version;
+use Neos\EventStore\Model\EventEnvelope;
+
+/**
+ * @internal CR upgrade internals
+ */
+final class EventEnvelopeFactory
```
Comment (bwaidelich) [inline Neos.ContentRepositoryRegistry/Classes/Upgrade/Shared/EventEnvelopeFactory.php:22]: I think it's fine to add some `EventEnvelopeFactory` but why is it in the `Upgrade` sub namespace?

**Review by bwaidelich — COMMENTED** [review]
I'm currently not able to seriously evaluate the `EventsDuplicateContentStreamRemovalUpgrade` logic (even though you did a great job keeping the code clean and even adding tests).
So for now just some not so helpful comments

File: Neos.ContentRepository.Export/Tests/Behavior/Features/Bootstrap/FeatureContext.php (hunk 1 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Export/Tests/Behavior/Features/Bootstrap/FeatureContext.php]: IDEs 😂

File: Neos.ContentRepositoryRegistry/Classes/Upgrade/Shared/EventEnvelopeFactory.php (hunk 3 above)
Comment (mhsdesign) [reply Neos.ContentRepositoryRegistry/Classes/Upgrade/Shared/EventEnvelopeFactory.php:22]: its definitely internal. It _should_ probably belong to the eventstore doctrine-dbal adapter and be there internal and here just sneakily used.
No other code aside from the eventstore and the upgrades will need to construct event envelopes themselves from the doctrine dbal specific table row format (see fromArray)

File: Neos.ContentRepositoryRegistry/Classes/Upgrade/Command/CRUpgradeCommandController.php (hunk 2 above)
Comment (mhsdesign) [reply Neos.ContentRepositoryRegistry/Classes/Upgrade/Command/CRUpgradeCommandController.php]: jup uuups ... if i describe what im doing here i hope it will be better understood as well:)

File: Neos.ContentRepositoryRegistry/Classes/Upgrade/EventsDeduplicateBaseWorkspaceChanges/EventsDeduplicateBaseWorkspaceChangesUpgrade.php [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+        $sequenceNumbersToRemove = [];
+
+        foreach ($duplicateContentStreamRemovalWithStreams as $duplicateContentStreamRemovals) {
+            $stream = StreamName::fromString($duplicateContentStreamRemovals['stream']);
+
+            if (!ContentStreamEventStreamName::isContentStreamStreamName($stream)) {
+                $this->log(sprintf('Error found illegal content stream removal event on non content stream %s', $stream->value));
+            }
+
```
Comment (nezaniel) [inline Neos.ContentRepositoryRegistry/Classes/Upgrade/EventsDeduplicateBaseWorkspaceChanges/EventsDeduplicateBaseWorkspaceChangesUpgrade.php:88]: haven't we done the following in line 51 already?

File: Neos.ContentRepositoryRegistry/Classes/Upgrade/EventsDeduplicateBaseWorkspaceChanges/EventsDeduplicateBaseWorkspaceChangesUpgrade.php [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+                if (!str_starts_with($correlationId->value, 'ChangeBaseWorkspace_')) {
+                    $this->log(sprintf('Error resolve duplicate content stream removal of %s as it was not caused due to a ChangeBaseWorkspace', $duplicateContentStreamRemovals['stream']));
+                    return;
+                }
+            }
+
+            $sequenceNumbers = array_map(intval(...), explode(',', $duplicateContentStreamRemovals['sequenceNumbers']));
+            $lowestSequenceNumber = SequenceNumber::fromInteger(min($sequenceNumbers));
+            $highestSequenceNumber = SequenceNumber::fromInteger(max($sequenceNumbers));
+
```
Comment (nezaniel) [inline Neos.ContentRepositoryRegistry/Classes/Upgrade/EventsDeduplicateBaseWorkspaceChanges/EventsDeduplicateBaseWorkspaceChangesUpgrade.php:105]: At this point, what additional information is required? We already know the duplicates, why not just remove them and be done with it?

**Review by nezaniel — CHANGES_REQUESTED** [review]

File: Neos.ContentRepositoryRegistry/Classes/Upgrade/EventsDeduplicateBaseWorkspaceChanges/EventsDeduplicateBaseWorkspaceChangesUpgrade.php (hunk 5 above)
Comment (mhsdesign) [reply Neos.ContentRepositoryRegistry/Classes/Upgrade/EventsDeduplicateBaseWorkspaceChanges/EventsDeduplicateBaseWorkspaceChangesUpgrade.php:105]: we only have the duplicate content stream removals. We cannot just eradicate these. To be correct we have to eradicate the whole base workspace change sequence. (see tests we expect 3 events less on the stream)

File: Neos.ContentRepositoryRegistry/Classes/Upgrade/EventsDeduplicateBaseWorkspaceChanges/EventsDeduplicateBaseWorkspaceChangesUpgrade.php (hunk 4 above)
Comment (mhsdesign) [reply Neos.ContentRepositoryRegistry/Classes/Upgrade/EventsDeduplicateBaseWorkspaceChanges/EventsDeduplicateBaseWorkspaceChangesUpgrade.php:88]: there we only validated the correlation id ... this is a sanity check for the content stream ... a was removed event must never be published on a workspace stream ... just one more check ... should never fail but better late than sorry

Comment (mhsdesign) [reply Neos.ContentRepositoryRegistry/Classes/Upgrade/EventsDeduplicateBaseWorkspaceChanges/EventsDeduplicateBaseWorkspaceChangesUpgrade.php:88]: wait jumped in line yes. Removed.

**Comment by mhsdesign** [comment]
Ill merge this later this evening or tomorrow as its required for the Neos 9.2 beta - and there are all tests for hopefully all edge cases - just added one more :) ... and what does not seem right will not be touched ... there are a lot of sanity checks.
